### PR TITLE
AIs can now cancel bolting/shocking. Bartenders can stop shaking their shaker.

### PR DESCRIPTION
**Pull request by Contrabang**
## What Does This PR Do

AIs can now cancel bolting/shocking by clicking on that door again (ctrl click will cancel door bolting, etc.). Bartenders can use their shaker while its being shook to cancel shaking it. Also adds a bit of features to do_after_once to allow for multiple do_after_onces on a single target, allowing you to bolt/shock the door at the same time. The bartender's shaker also has a 3 second cooldown if you cancel it, so you cant just spam the sound/text any more than you used to be able to

## Why It's Good For The Game
These are two things that are allow_moving do_afters, that have no clear way to cancel them. For the AI, you cant cancel at all. For the bartender, you can put the shaker in your backpack/drop it, but thats clunky.

## Images of changes

image
image

## Testing

Started bolting door, canceled bolting door.
Started bolting door, waited for it to bolt. Bolted.
Started shocking door, canceled shocking door.
Started shocking door, waited for it to shock. Shocked.
Started bolting and shocking at the same time, canceled both.
Started bolting and shocking at the same time, waited, door was both shocked and bolted.

Started shaking a drink, canceled it with attacking myself
Started shaking a drink, waited for it to shake. It was mixed.

## Changelog
:cl:
tweak: AIs can now cancel their bolting/shocking ctrl or alt clicking the door again.
tweak: The bartender's shaker can now stop shaking by attacking yourself again.
/:cl:

### CODE REVIEW

**Comment by go-liath-online** [comment]
Concerned about this just resulting in malf testing. A la "AI if you can bolt and unbolt this faster than once every four seconds, you're malf."

**Comment by Contrabang** [comment]
> Concerned about this just resulting in malf testing. A la "AI if you can bolt and unbolt this faster than once every four seconds, you're malf."

They can technically already bolt/electrify faster if they're willing to open the menu. Thats more just an issue with the timer being on bolting/shocking overall IMO. I can try to change it if its a major concern for design/balance team

File: code/modules/food_and_drinks/drinks/drinks_base.dm [hunk 1]
```diff
@@ -288,21 +290,22 @@
 		to_chat(user, "<span class='warning'>You won't shake an empty shaker now, will you?</span>")
 		return
 
-	if(shaking)
-		return
+	if(COOLDOWN_FINISHED(src, shaking_cooldown))
+		shaking = TRUE
+		var/adjective = pick("furiously", "passionately", "with vigor", "with determination", "like a devil", "with care and love", "like there is no tomorrow")
+		user.visible_message("<span class='notice'>\The [user] shakes \the [src] [adjective]!</span>", "<span class='notice'>You shake \the [src] [adjective]!</span>")
```
Comment (lewcc) [inline code/modules/food_and_drinks/drinks/drinks_base.dm:296]:
```suggestion
		user.visible_message("<span class='notice'>[user] shakes [src] [adjective]!</span>", "<span class='notice'>You shake [src] [adjective]!</span>")
```

File: code/modules/food_and_drinks/drinks/drinks_base.dm [hunk 2]
```diff
@@ -288,21 +290,22 @@
...
+		icon_state = "shaker-shake"
+		if(iscarbon(loc))
+			var/mob/living/carbon/M = loc
+			M.update_inv_r_hand()
+			M.update_inv_l_hand()
+		playsound(user, 'sound/items/boston_shaker.ogg', 80, 1)
```
Comment (Henri215) [inline code/modules/food_and_drinks/drinks/drinks_base.dm:302]:
```suggestion
		playsound(user, 'sound/items/boston_shaker.ogg', 80, TRUE)
```

**Review by Henri215 — APPROVED** [review]

**Review by S34NW — APPROVED** [review]

**Pull request merged**